Adds support for docker-machine

labianchin opened the pull request:
Related to issue #50.
Although it is already working I'd still consider this as a preview.
It seems to be missing to update Readme, add tests and improve the install (e.g. add_environment_variables is not important for docker-machine).
Can you please provide some feedback? And also what would you expect to get this merged?

labianchin committed:
Adds support for docker-machine

labianchin committed:
Add docker-machine support to readme.

Fix port parameter in do_rsync
Install will add to host based on docker-machine name and ip
Install does not set the environment when using docker-machine

brikis98 commented:
Thanks for the pull request. I apologize for the delay, but I'm currently on the road with limited Internet access, so I'll take a look once I'm back (should be early August).

ilkka commented:
Since probably in almost all cases docker-osx-dev would be run against the active machine, I think you could drop the requirement to set the machine name environment variable on every invocation, and instead just use `$(docker-machine active)` to get it. Thoughts?

labianchin commented:
@brikis98, no problem, I still have to add some tests.

@ilkka, sorry getting ahead of you, but I needed it this week.
Correct, but there is the case where you can have multiple active docker-machines. What about passing the machine name as first argument?

```
docker-osx-dev machine-name
```

Also, I am trying to maintain compatibility with boot2docker. Later if needed we can drop support for it.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -379,13 +384,63 @@ function init_boot2docker {
+################################################################################
+
+function get_json_val () { 
+    python -c "import json,sys;sys.stdout.write(json.dumps(json.load(sys.stdin)$1))" | sed -e 's/^"//'  -e 's/"$//'
```
I'm a bit concerned about adding a dependency on python. Any alternative way to do this?

labianchin replied:
Agreed! That is a hard one and I will have to think better on this. I saw it is possible to do with sed, grep and awk. 
Some examples that I saw using grep actually use GNU grep, which is not always available on OSX. I am open to any suggestion.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -379,13 +384,63 @@ function init_boot2docker {
+# docker-machine manipulation
+################################################################################
+
+function get_json_val () { 
```
This function needs a better name and comments explaining how it is intended to be used. As it is, I can't tell what JSON it's reading (from a file or stdin or as a param?) nor what "val" it's trying to get (a single key or the whole JSON blob?).

labianchin replied:
Agreed. I will change to `get_json_value` and add some docs, but I still have work in the python dependency.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -379,13 +384,63 @@ function init_boot2docker {
+  DOCKER_MACHINE_STORE_PATH=$(echo -e "$DOCKER_MACHINE_INSPECT" | get_json_val "['StorePath']")
+
+  DOCKER_HOST_SSH_URL="$DOCKER_HOST_USER@$DOCKER_HOST_IP"
+  DOCKER_HOST_SSH_KEY="$DOCKER_MACHINE_STORE_PATH/id_rsa"
```
Any way to avoid hard-coding `id_rsa`?

labianchin replied:
I couldn't find any way to do that. `docker-machine inspect` doesn't expose this.

brikis98 commented on README.md:
````diff
@@ -144,6 +144,21 @@ http://dockerhost:5000
+
+```
+> cd /foo/bar
+> DOCKER_MACHINE_NAME=foo docker-osx-dev
````
Any reason to use an environment variable here instead of a flag? It seems more consistent with the docker-osx-dev design to do something like:

```
docker-osx-dev --machine-name foo
```

labianchin replied:
No specific reason. I think I just hadn't thought about it :) . Will do the change.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -672,6 +729,11 @@ function sync {
 
+function docker_host_ssh {
+  docker-machine ssh "$DOCKER_MACHINE_NAME" "$@"
+  # TODO: or boot2docker ssh
```
Yea... :)

labianchin replied:
:+1:

brikis98 commented on src/docker-osx-dev:
```diff
@@ -971,10 +1032,10 @@ function watch_and_sync {
+  install_rsync_on_docker_host
   add_docker_host
-  add_environment_variables
+  [ -z "$DOCKER_MACHINE_NAME" ] && add_boot2docker_environment_variables
```
Probably better to create a function called `add_environment_variables` that takes care of the `$DOCKER_MACHINE_NAME` check itself.

labianchin replied:
Yes, will do that.

brikis98 commented:
Thank you for the PR. This is a pretty massive change that forks the code in quite a few places. Do we know for sure that using boot2docker directly via the CLI is deprecated? Is docker-machine definitely the way to interact with Docker on the desktop in the future? If so, it's worth thinking about supporting _only_ docker-machine. This project is hard to test (see #7), so I worry about any complications that fork the code too much.

bgentry commented:
The boot2docker-cli deprecation notice is here: https://github.com/boot2docker/boot2docker-cli#pending-deprecation

I have used docker-machine locally on OS X and it seems to work just fine, not sure if there's any reason to keep supporting the boot2docker-cli instead. Maybe somebody from Docker or somebody that has more knowledge around that can chime in.

brikis98 commented:
Hmm, the [Docker Docs](https://docs.docker.com/installation/mac/) still point to boot2docker and using it as a command line tool.

labianchin committed:
Fixes from the feedback of brikis98

- pass --machine-name as argument, instead of env var
- ssh command working also for boot2docker ssh
- add_docker_environment_variables function that adds env vars only for
boot2docker
- add doc and rename get_json_value

labianchin commented:
Thanks @brikis98 for the feedback!

I made some changes but, for me, there are at least two things outstanding:
- Find an alternative to using python to parse `docker-machine inspect` json.
- Add some automated tests

Regarding the boot2docker deprecation, my understanding is that it is slowly being deprecated. I'd suggest for now keeping support for both `boot2docker` and `docker-machine`. But in the near future, `boot2docker` support should be dropped.

brikis98 commented:
OK, let's keep the boot2docker support a bit longer. Let me know if you have ideas on the JSON & automated test front.

labianchin committed:
Adds tests to check proper DOCKER_HOST variables configuration

ilkka commented:
> Correct, but there is the case where you can have multiple active docker-machines.

@labianchin, multiple active? I thought the active machine was determined by DOCKER_HOST in the env, meaning there can only be one?

ilkka commented:
just tried this branch again and noticed a weird thing: my freshly generated virtualbox docker machine has a config like this:

```
{
    "DriverName": "virtualbox",
    "Driver": {
        "IPAddress": "192.168.99.100",
        ...
        "SSHPort": 55045,
        ...
    }
   ...
}
```

which makes `docker-osx-dev sync` try to connect to port 55045, which fails. `docker-machine ssh <machine>` works just fine though, as does manually ssh'ing to port 22. I have no idea why that value is there or what it's significance is, but set up like this, the sync doesn't work.

ilkka commented:
Ah, now I understand what is happening with `SSHPort`: that's the port on _localhost_ that the vbox SSH port is forwarded to. So if you're gonna use SSHPort for rsync, you should connect to localhost, not the docker machine itself.

labianchin commented:
> @labianchin, multiple active? I thought the active machine was determined by DOCKER_HOST in the env, meaning there can only be one?

@ilkka, Yes. That makes more sense when you use a driver like EC2, OpenStack or GCE. This way you can have multiple docker-machine active (and running) in the cloud. You might want to have a docker-machine per each project you have.

So in one terminal tab you can run:

```
cd project-foo && docker-osx-dev --machine-name machine-for-project-foo
```

And in the other:

```
cd project-bar && docker-osx-dev --machine-name machine-for-project-bar
```

labianchin commented:
> which makes docker-osx-dev sync try to connect to port 55045, which fails. docker-machine ssh <machine> works just fine though, as does manually ssh'ing to port 22. I have no idea why that value is there or what it's significance is, but set up like this, the sync doesn't work.

@ilkka Thanks for catching that! I was testing more with the openstack driver. There the `SSHPort` was 22. I will take a look on that. I guess I will have to understand better which params `docker-machine ssh` use.

labianchin committed:
Remove port configuration on rsync, try to mimic docker-machine scp
arguments

labianchin commented:
@ilkka 

So I was looking into what is the arguments `docker-machine scp` use, with the virtualbox driver it is this:

```
/usr/bin/scp [/usr/bin/scp -o IdentitiesOnly=yes -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=quiet -3 -i /Users/me/.docker/machine/machines/local/id_rsa foo.txt docker@192.168.99.100:/]
```

So, I just removed the `-p` argument for the `rsync --rsh`. Now should be fine.

But, one problem I am getting is that even after `install_rsync_on_docker_host`, and with rsync on `/usr/local/bin` it keeps getting `-sh: rsync: not found`.

It seems rsync 64bits is now available for `tce-load`: https://github.com/boot2docker/boot2docker/issues/936#issuecomment-120992953. @brikis98, should we use it?

ilkka commented:
> You might want to have a docker-machine per each project you have.

@labianchin ah yes, but since active is stored in the env, it is per terminal session, so you can run multiple machines that way too, just in separate terminals. Essentially it's the same. I don't mind having the parameter, I just think that defaulting to whatever `docker-machine active` outputs saves typing and is in no way surprising behaviour :)

> But, one problem I am getting is that even after install_rsync_on_docker_host, and with rsync on /usr/local/bin it keeps getting -sh: rsync: not found.

I found that I had to `docker-osx-dev install` before it found rsync after restarting (but not recreating) my docker machine, although I might have been hallucinating from all the coffee 😅 Don't know if it's related at all.

brikis98 commented:
@labianchin: Yes, if the 64 bit rsync is now working, we should go back to using `tce-load` to install it. I think the original code is still in the comments. I've been meaning to do this, but haven't had a chance to get to it, so PRs are welcome :)

labianchin committed:
Remove python dependecy python to get json values from docker-machine inspect

labianchin commented:
@ilkka, that make sense. I agree that would be a good default. But if, for now, we are keeping the support for boot2docker, then it would be better to keep the default (no `--machine-name` arg) as using boot2docker.

Also, I just run `docker-machine active` and I get this:

```
dev
```

But when I run `docker-machine ls`:

```
NAME         ACTIVE   DRIVER       STATE     URL                         SWARM
dev                   virtualbox   Stopped
dev1                  virtualbox   Stopped
dokku-test            openstack    Running   tcp://10.254.132.163:2376
lbianchin2            openstack    Running   tcp://10.254.132.14:2376
local                 virtualbox   Running   tcp://192.168.99.102:2376
```

That is weird.

labianchin commented:
So, I added some test cases and removed the python dependency. @brikis98 can you take a look?

I also rebased on top of master.

I run some simple manual tests and things look good. What I did was the following:

```
docker-machine -D create --driver virtualbox local
docker-osx-dev --machine-name local -e .git -l DEBUG install
docker-osx-dev --machine-name local -e .git -l DEBUG
touch foo
```

brikis98 commented on README.md:
````diff
@@ -144,6 +144,21 @@ http://dockerhost:5000
+[INFO] Watching: /foo/bar
+```
+
+Note: currently this assumes the docker-machine was already created.
````
Instead of this note, could you include the `docker-machine create` command in the example above, similar to how you did in [this comment](https://github.com/brikis98/docker-osx-dev/pull/77#issuecomment-128207268)?

labianchin replied:
:+1:

brikis98 commented on src/docker-osx-dev:
```diff
@@ -375,17 +381,74 @@ function init_boot2docker {
+################################################################################
+
+#
+# Read a json string from stdin and return the value of the key provided as first argument
```
Can you add an example call to this documentation? It makes it more obvious what the method does.

labianchin replied:
:+1:

brikis98 commented on src/docker-osx-dev:
```diff
@@ -375,17 +381,74 @@ function init_boot2docker {
+# Configures variables based on a the docker-machine inspect from $DOCKER_MACHINE_NAME
+#
+function configure_docker_machine {
+  DOCKER_HOST_NAME=$DOCKER_MACHINE_NAME
```
`$DOCKER_MACHINE_NAME` in quotes?

labianchin replied:
:+1:

brikis98 commented on src/docker-osx-dev:
```diff
@@ -375,17 +381,74 @@ function init_boot2docker {
+}
+
+#
+# Configures variables based on a the docker-machine inspect from $DOCKER_MACHINE_NAME
```
Minor typo: "a the"

labianchin replied:
:+1:

brikis98 commented on src/docker-osx-dev:
```diff
@@ -375,17 +381,74 @@ function init_boot2docker {
+  log_info "Installing rsync in the Docker Host image"
 
-  boot2docker ssh "if ! type rsync > /dev/null 2>&1; then tce-load -wi rsync; fi"
+  $DOCKER_HOST_SSH_COMMAND "if ! type rsync > /dev/null 2>&1; then tce-load -wi rsync; fi"
```
Quick sanity check, in case you haven't done it already: search the code for `boot2docker ssh` calls and make sure all of them are updated to `$DOCKER_HOST_SSH_COMMAND`.

labianchin replied:
:+1: Yes, there are two left that I forgot. Thanks for calling that out.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -428,16 +491,18 @@ function get_env_file {
-    log_instructions "To pick up important new environment variables in the current shell, run:\n\tsource $env_file"
-  else
-    log_warn "All Boot2Docker environment variables already defined, will not overwrite"
+  if [ -z "$DOCKER_MACHINE_NAME" ]; then
```
I haven't used docker-machine much, so just to confirm: it doesn't rely on _any_ environment variables?

labianchin replied:
Good catch. Well, it does. Please see here: https://github.com/brikis98/docker-osx-dev/pull/77/files#r36412414

brikis98 replied:
So should the call that adds those env vars be moved into this function instead?

labianchin replied:
We could. But `add_environment_variables` is writing to `bashrc` or `zshrc`. `docker-machine` env vars should be avoided to be written there. If the user has multiple machines the env vars would get duplicated.
Also the purpose of the `init_docker_machine` is to run the commands from the machine documentation: `start` and `env`.

brikis98 replied:
Ah, gotcha.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -657,6 +723,11 @@ function sync {
   done
 }
 
+function docker_host_ssh {
```
What's this?

labianchin replied:
I forgot this, I am removing.

brikis98 commented on src/docker-osx-dev:
```diff
@@ -726,6 +797,7 @@ function instructions {
   echo -e "  $INSTALL_COMMAND\tInstall docker-osx-dev and all of its dependencies."
   echo -e
   echo -e "Options:"
+  echo -e "  -n, --machine-name name\t\tWhen suplied syncs with the given docker machine host"
```
Why `-n` and not `-m`?

labianchin replied:
`-n` was supposed to mean _name_, but I guess `-m` is more clear.

brikis98 commented on test/docker-osx-dev.bats:
```diff
@@ -415,3 +415,45 @@ export NEW_ENV_VARIABLE_2=VALUE2"
   assert_output '/some/path'
 }
+
+@test "init_docker_host should call configure_boot2docker set DOCKER_HOST vars" {
```
You should add a bunch of unit tests for the `get_json_value` function.

labianchin replied:
:+1:

brikis98 commented:
@labianchin: Thanks! I just left a bunch more comments. Almost there.

ilkka commented:
@labianchin that `docker-machine active` behaviour is due to the _slightly_ different ways `active` and `ls` determine the active host. `active` _only_ looks at the value of DOCKER_HOST, but `ls` also checks that the host is running.

labianchin commented on src/docker-osx-dev:
```diff
@@ -375,17 +381,74 @@ function init_boot2docker {
+  docker-machine start "$DOCKER_MACHINE_NAME"
+  eval $(docker-machine env "$DOCKER_MACHINE_NAME")
+  configure_docker_machine
+}
```
So, here, this eval will run something like this:

```
export DOCKER_TLS_VERIFY="1"
export DOCKER_HOST="tcp://192.168.99.103:2376"
export DOCKER_CERT_PATH="/Users/someone/.docker/machine/machines/local"
export DOCKER_MACHINE_NAME="local"
```

We can see that `DOCKER_MACHINE_NAME` will be redefined.

Also, as in the `docker-machine` documentation, if the user runs `eval "$(docker-machine env <machine-name>)"`, it will define the `DOCKER_MACHINE_NAME` to the env. Then if he runs `docker-osx-dev`, he would not need to pass the `--machine-name` argument.

I guess this is fine, I can add to the `Readme.md` about this behavior.

ilkka replied:
:+1:

labianchin commented:
@brikis98 thanks for the feedback, I will be making the fixes.

@ilkka Please see this: https://github.com/brikis98/docker-osx-dev/pull/77#discussion_r36412414

So running `eval "$(docker-machine env <machine-name>)"` would mean the user does not need use the `--machine-name` argument.

nathanleclaire commented:
One idea I've tinkered with is some kind of `--delta` flag to `docker-machine scp` which would shell out to `rsync` if available instead of `scp`.  Would that be of interest to you / help your workload on this project?

labianchin commented:
Yes @nathanleclaire, I would be interested! Can you provide some details on what you have in mind? Probably via e-mail. Thanks!

labianchin committed:
Fixes in docs and typos, add more tests to get_json_value

labianchin commented:
@brikis98 I think I got all the suggested fixes in this last commit. Can you take a look?

brikis98 commented on src/docker-osx-dev:
```diff
@@ -375,17 +381,79 @@ function init_boot2docker {
+
+#
+# Read a json string from stdin and return the value of the key provided as first argument
+# Uses head -n 1 to avoid duplicates
```
This is an implementation detail. Instead, you should probably say something along the lines of "Does a simple string search and returns the first string value that matches the given key".

labianchin replied:
:+1:

brikis98 commented on src/docker-osx-dev:
```diff
@@ -942,9 +1014,7 @@ function configure_includes {
-  configure_boot2docker_ssh_key
-  init_boot2docker
-  install_rsync_on_boot2docker
+  init_docker_host
```
Why do you not call `install_rsync_on_docker_host` here as before?

labianchin replied:
I can move back here. But I believe this slows down the startup of `docker-osx-dev` after the install.

brikis98 replied:
It shouldn't cause much slow down, as the command checks if `rsync` is installed before trying to install it. The benefit is that if you create a new docker VM for whatever reason, rsync will get installed on it any time you run docker-osx-dev.

labianchin replied:
:+1:  I can add that back.

brikis98 commented on test/docker-osx-dev.bats:
```diff
@@ -415,3 +416,69 @@ export NEW_ENV_VARIABLE_2=VALUE2"
+  assert_output "first"
+}
+
+@test "get_json should not get value for a json for a non existent key" {
```
What is the expected behavior of `get_json` if the key does exist, but its value is not a string, but an integer, object, or array?

labianchin replied:
Yes, adding tests for that.

brikis98 commented:
@labianchin: A few more questions/comments above :)

labianchin committed:
Adds more tests for get_json_value and clarify docs

labianchin commented:
Hey @brikis98, I've made some fixes. Please also see my previous comments.

brikis98 commented on README.md:
````diff
@@ -144,6 +144,28 @@ http://dockerhost:5000
+```
+> docker-machine create --driver virtualbox <machine-name>
+> eval "$(docker-machine env <machine-name>)"
+> docker-osx-dev install
````
Sorry, just realized I have one more question on this: do you have to run the `install` command every time you run `docker-machine create`? Or only the `eval` part?

labianchin replied:
So, I am little confused on what should be the role of the `install` for `docker-machine`.

``` bash
   install_dependencies # Should actually run before docker-machine create. Once for osx user.
   init_docker_host # Per machine, should run after docker-machine create. This seems only important for the next command
   install_rsync_on_docker_host # Is this really important? the sync will also install rsync.
   add_docker_host # Per machine, should run after docker-machine create. Also, is it important?
   add_environment_variables # Not used for docker-machine
```

I would avoid making the `install` create the docker-machine, as in the creation the user needs to supply the driver he wants and all the driver's configuration. It is also seems misleading that a `install` command would create a machine.

I'd suggest the following flow:

```
docker-osx-dev install_dependencies # once per osx user
docker-machine create --driver virtualbox <machine-name>
eval "$(docker-machine env <machine-name>)"
cd /foo/bar
docker-osx-dev
```

In this way there is no need to run `install` after `docker-machine create`. The `install` would only make sense for `boot2docker`.

labianchin committed:
Adds back install_rsync_on_docker_host to watch_and_sync

brikis98 commented:
I expect developers to run the `docker-osx-dev` script very frequently, so we should keep its startup time well under one second in the normal case. Therefore, the `install` command should take care of anything that: 
1. Takes a long time (more than ~1 second)
2. Anything that needs to happen only once per computer
3. Anything that needs to be run before you can use docker (i.e., install docker itself)

This is why it includes `install_dependencies`, `init_docker_host`, `install_rsync_on_docker_host`, `add_docker_host`, and `add_environment_variables`. The actual `docker-osx-dev` script, in theory, should run _just_ the `watch_and_sync` command. However, it runs a few other steps (e.g., `install_rsync_no_docker_host`) that: 
1. _Might_ go out of sync as the user tweaks their docker environment
2. Are very fast to run in the normal case (e.g., checking if rsync is installed is very quick)

The question I asked earlier, though I guess not very clearly, was whether the `install` command should automate the `eval "$(docker-machine env <machine-name>)"` command you have listed in the README, or any other manual steps necessary for using docker-machine?

labianchin commented:
Yes. That really make sense. Thanks for clarifying.

Answering your question: I believe it shouldn't, as users might want to switch between different docker-machine instances (machines). So every time they run `docker-osx-dev` (`watch_and_sync`), they supply which machine name they want to run with.

So, clarifying, this steps, should run just after creating a new machine instance:

``` bash
eval "$(docker-machine env <machine-name>)"
docker-osx-dev install
```

Which is equivalent of:

``` bash
docker-osx-dev --machine-name <machine-name> install
```

After the instance is created and `docker-osx-dev` installed, users only need to run:

``` bash
eval "$(docker-machine env <machine-name>)"
docker-osx-dev
```

Or the equivalent:

``` bash
docker-osx-dev --machine-name <machine-name>
```

brikis98 commented:
Awesome, thanks for the PR. It's now merged.

labianchin commented:
Thanks for all the feedback! That has been a great experience contributing to this project.
